backend/src/common/device/keypad.rs

```diff
@@ -0,0 +1,283 @@
+            capacity,
+        }
+    }
+    pub fn add_keys(&mut self, keys: HashSet<char>) {
+        if !self.is_pressed_keys_initialized {
+            self.is_pressed_keys_initialized = true;
+            self.previous_pressed_keys = keys;
+            return;
+        }
+        let unpressed_keys = self.previous_pressed_keys.difference(&keys);
+        for ch in unpressed_keys {
+            self.data.push_back(ch.clone());
+            if self.data.len() == self.capacity {
+                self.data.pop_front();
+            }
+        }
+        self.previous_pressed_keys = keys;
+    }
+    pub fn get_candidate_keys(&mut self) -> Vec<String> {
+        let candidates = self
+            .data
+            .iter()
+            .collect::<String>()
+            .split_inclusive(CandidateKey::ENDING_KEY)
+            .filter(|x| x.ends_with(CandidateKey::ENDING_KEY))
+            .map(|x| x.trim_end_matches(CandidateKey::ENDING_KEY))
+            .map(|x| x.to_owned())
+            .collect::<Vec<String>>();
+        let length: usize = candidates.iter().map(|x| x.len() + 1).sum();
+        for _ in 0..length {
+            self.data.pop_front();
+        }
+        candidates
+    }
+    pub fn clear(&mut self) {
+        self.data.clear();
+        self.previous_pressed_keys.clear();
+        self.is_pressed_keys_initialized = false;
+    }
+}
+
+#[derive(Clone)]
+pub struct KeyPad {
+    code: Code,
+    matrix: KeyPadMatrix,
+    potential_key: CandidateKey,
+    last_pressed: Instant,
+}
+
+impl KeyPad {
+    pub const RESET_TIMER: Duration = Duration::from_secs(5);
+    pub fn new(
+        code: Code,
+        matrix: KeyPadMatrix,
+        potential_key: CandidateKey,
+        last_pressed: Instant,
+    ) -> KeyPad {
+        KeyPad {
+            code,
+            matrix,
+            potential_key,
+            last_pressed,
+        }
+    }
+    pub fn add_keys(&mut self) {
+        let keys = self.matrix.get_keys_pressed();
+        if keys.len() != 0 {
+            self.last_pressed = Instant::now();
+        }
+        self.potential_key.add_keys(keys);
+    }
+    pub fn check_candidates(&mut self) -> bool {
+        let candidates = self.potential_key.get_candidate_keys();
+        candidates
+            .iter()
+            .rev()
+            .take(1)
+            .map(|x| self.code.is_candidate_valid(x))
+            .all(|x| x)
+    }
+    pub fn get_last_pressed(&self) -> Instant {
+        self.last_pressed
+    }
+    pub fn reset_input_keys(&mut self) {
+        self.potential_key.clear();
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Code {
+    pub data: String,
+}
+
+impl Code {
+    const VALID_CHARS: [char; 14] = [
+        'A', 'B', 'C', 'D', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0',
+    ];
+    const START_UP_CODE: &'static str = "0000";
+    fn validate_string(string: &str) -> bool {
+        string
+            .chars()
```

Comment:
are we sticking with the 4 digit code? if so can we add a validation for it

Reply:
we're not.